common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/SmartcardNfcCertBasedAuthChallengeHandler.java

```diff
@@ -0,0 +1,175 @@
+            @Override
+            public void onCreateConnection() {
+                //Reset DialogHolder to null if necessary.
+                //In this case, DialogHolder would be an ErrorDialog if not null.
+                mDialogHolder.dismissDialog();
+            }
+
+            @Override
+            public void onClosedConnection() {
+                //Nothing needed.
+            }
+        });
+    }
+
+    /**
+     * Additional logic to run upon before the end of processChallenge.
+     */
+    @Override
+    protected void onGetCertSessionExtendedLogic() {
+        mSmartcardCertBasedAuthManager.stopDiscovery(mActivity);
+    }
+
+    /**
+     * Upon a positive button click in the smartcard PIN dialog, verify the provided PIN and handle the results.
+     * @param certDetails ICertDetails of the selected certificate from the SmartcardCertPickerDialog.
+     * @param request ClientCertRequest received from AzureActiveDirectoryWebViewClient.onReceivedClientCertRequest.
+     * @return A PositiveButtonListener to be set for a SmartcardPinDialog.
+     */
+    @Override
+    protected SmartcardPinDialog.PositiveButtonListener getSmartcardPinDialogPositiveButtonListener(@NonNull final ICertDetails certDetails,
+                                                                                                    @NonNull final ClientCertRequest request) {
+        final String methodTag = TAG + ":getSmartcardPinDialogPositiveButtonListener";
+
+        return new SmartcardPinDialog.PositiveButtonListener() {
+            @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+            @Override
+            public void onClick(@NonNull final char[] pin) {
+                //For NFC, we need another dialog prompting the user to hold the smartcard to the phone again.
+                mDialogHolder.showSmartcardNfcPromptDialog(new SmartcardNfcPromptDialog.CancelCbaCallback() {
+                    @Override
+                    public void onCancel() {
+                        mDialogHolder.dismissDialog();
+                        mTelemetryHelper.setResultFailure(USER_CANCEL_MESSAGE);
+                        request.cancel();
+                    }
+                });
+                mSmartcardCertBasedAuthManager.setConnectionCallback(new AbstractSmartcardCertBasedAuthManager.IConnectionCallback() {
+                    @Override
+                    public void onCreateConnection() {
+                        mDialogHolder.showDialog(new SmartcardNfcLoadingDialog(mActivity));
+                        mSmartcardCertBasedAuthManager.requestDeviceSession(new AbstractSmartcardCertBasedAuthManager.ISessionCallback() {
+                            @Override
+                            public void onGetSession(@NonNull final ISmartcardSession session) throws Exception {
+                                tryUsingSmartcardWithPin(pin, certDetails, request, session);
```

Comment:
what would happen if I tap with a different key?

Reply:
I'm wondering if it's possible to front-load pin prompt (even before entering NFC/USB flow)

Reply:
:^) good point. added a method isCurrDiffFromPrevConnectedDevice() to abstract class AbstractNfcSmartcardCertBasedAuthManager, and this method is called before  tryUsingSmartcardWithPin() is called.
If it returns true, then a generic error dialog will show and the flow will cancel. Showing a generic error dialog for now since creating a more specific one requires the designer and localization, so that will be replaced in a later PR.